File: packages/java/anip-server/src/main/java/dev/anip/server/ApprovalDecisionResult.java
package dev.anip.server;

import dev.anip.core.ApprovalGrant;

/**
 * Discriminated result for {@link Storage#approveRequestAndStoreGrant}.
 *
 * <p>{@code ok=true} -> {@link #grant} is set, {@link #reason} is null.
 * {@code ok=false} -> {@link #reason} is one of:
 * {@code approval_request_not_found}, {@code approval_request_expired},
 * {@code approval_request_already_decided}; {@link #grant} is null.</p>
 *
 * <p>v0.23. See SPEC.md §4.9 (Decision 0.9a — atomic approve+issue).</p>
 */
public record ApprovalDecisionResult(boolean ok, ApprovalGrant grant, String reason) {

    public static ApprovalDecisionResult success(ApprovalGrant grant) {
        return new ApprovalDecisionResult(true, grant, null);
    }

    public static ApprovalDecisionResult failure(String reason) {
        return new ApprovalDecisionResult(false, null, reason);
    }
}

File: packages/csharp/src/Anip.Server/ApprovalDecisionResult.cs
using Anip.Core;

namespace Anip.Server;

/// <summary>
/// Result of <see cref="IStorage.ApproveRequestAndStoreGrantAsync"/>. v0.23.
/// SPEC.md §4.9 (Decision 0.9a) — atomic state transition for an
/// ApprovalRequest into "approved" + grant insert.
/// On failure, <see cref="Reason"/> is one of:
/// <list type="bullet">
///   <item>approval_request_not_found</item>
///   <item>approval_request_already_decided</item>
///   <item>approval_request_expired</item>
/// </list>
/// </summary>
public record ApprovalDecisionResult(bool Ok, ApprovalGrant? Grant, string? Reason)
{
    public static ApprovalDecisionResult Success(ApprovalGrant grant) =>
        new(true, grant, null);

    public static ApprovalDecisionResult Failure(string reason) =>
        new(false, null, reason);
}
